Improving theme cloning process in order to minimize manual updates

shaal opened the pull request.

ccjjmartin commented on emulsify.php:
```diff
@@ -408,6 +408,8 @@ function _emulsify_get_files_to_alter() {
     'emulsify.theme',
     'emulsify.breakpoints.yml',
     'emulsify.libraries.yml',
+    'templates/navigation/menu--main.html.twig',
```
It looks like these two lines should be removed since they are already copied by default when the slim option is not used.  See a few lines down where we copy the entire contents of the `templates` and `components` directories.  This would add these files to the `--slim` option which is designed not to copy any files from these directories and be a bare bones install.  Let me know if you were having some trouble with these files not showing up as expected.

shaal replied:
@ccjjmartin, I was not aware of the `--slim` option.
When I run `php emulsify.php -slim myslimtheme` I am getting an error -
```
PHP Warning:  copy(/home/shaal/code/lando/rosenberg/web/themes/contrib/emulsify/components/css): failed to open stream: No such file or directory in /home/shaal/code/lando/rosenberg/web/themes/contrib/emulsify/emulsify.php on line 125
Failed on Phase: Copy files.Your theme was not successfully created.
```
I think the following line of code is trying to copy a directory that does not exist - `components/css`
https://github.com/fourkitchens/emulsify/blob/develop/emulsify.php#L495

shaal replied:
@ccjjmartin, I updated the code.
Now it doesn't try copying `components/css` directory
I moved the 2 files that need to be altered to the right place so it works as intended with the `-slim` option or without.
https://github.com/fourkitchens/emulsify/pull/268#commits-pushed-1c572e3

ccjjmartin commented on components/_patterns/01-atoms/04-images/icons/_icon.twig:
```diff
@@ -8,10 +8,9 @@
-{# For Drupal: Update this path to match your theme directory structure #}
 {# For Pattern Lab: The icon_url is defined in _data/paths.json. No action is required #}
-{% set icon_url = icon_url|default('/themes/custom/YOURTHEME/dist/img/sprite/') %}
-{# Update `emulsify` here to your theme's machine name #}
```
@shaal I am curious as to the broken behavior you experienced here.  Did "emulsify" on this line get replaced?  On the line right after this did it not get replaced?  I am definitely pro simplifying install processes so I would love to hear more about the bug you encountered.

ModulesUnraveled replied:
I think the goal here is to set this as `emulsify` by default so that it can be auto-replaced. In the "before" code, it says to manually enter your theme's machine name. Setting it automatically on cloned theme creation is preferred.

ccjjmartin replied:
Created this ticket for follow up on this: https://github.com/fourkitchens/emulsify/issues/271

shaal commented:
@ccjjmartin, I updated the code.
Now it doesn't try copying `components/css` directory
I moved the 2 files that need to be altered to the right place so it works as intended with the `-slim` option or without.
https://github.com/fourkitchens/emulsify/pull/268#commits-pushed-1c572e3

ccjjmartin commented on emulsify.php:
```diff
@@ -415,6 +415,8 @@ function _emulsify_get_files_to_alter() {
   }
   else {
     return array_merge($default_array, array(
+      'templates/navigation/menu--main.html.twig',
```
Can you remove these two lines completely and apply this patch to your branch:
```
diff --git a/emulsify.php b/emulsify.php
index a3bec1c..a889a5e 100644
--- a/emulsify.php
+++ b/emulsify.php
@@ -550,7 +550,7 @@ function _emulsify_get_files_to_rename() {
  * @return boolean
  *   A boolean representing the success or failure of the function.
  */
-function _emulsify_alter_files($theme_path, array $files_to_alter = array(), array $alterations = array(), $absolute = FALSE) {
+function _emulsify_alter_files($theme_path, array $files_to_alter = array(), array $alterations = array(), $absolute = FALSE, int $depth = 0) {
   if (empty($files_to_alter) || empty($alterations)) {
     return TRUE;
   }
@@ -568,11 +568,11 @@ function _emulsify_alter_files($theme_path, array $files_to_alter = array(), arr
       $files = scandir($file_path);
       $files = array_splice($files, 2);
       foreach ($files as $file) {
-        $processed_files[] = $file_path . $file;
-      }
-      $alter_status = _emulsify_alter_files($theme_path, $processed_files, $alterations, TRUE);
-      if ($alter_status === FALSE) {
-        return FALSE;
+        $processed_file = [$file_path . DIRECTORY_SEPARATOR . $file];
+        $alter_status = _emulsify_alter_files($theme_path, $processed_file, $alterations, TRUE, $depth + 1);
+        if ($alter_status === FALSE) {
+          return FALSE;
+        }
       }
     }
     elseif ($file_type === 'file') {
```

This was originally designed that you can just list the `components` directory and it recursively searches for files and replaces the text within them.  It must have broken somewhere along the way.  This fixes it though according to my tests, reproducing steps below:

1) Run `php emulsify.php "Alter Test"`
2) Verify that the _icon.twig file contains the name `alter_test` as the machine_name on the desired lines changed above.

NOTE: We may want to check to see if anything else unexpectedly broke since the replacements are now actually working throughout the components and templates directories.

ccjjmartin commented:
@shaal Also, there is another instance of `components/css` present in the file.  Can you remove that one too?

shaal commented:
@ccjjmartin, Thank you for the new code!
I committed the patch and removed the 2 lines that are no longer needed.
It seems to be working well on my machine.

regarding https://github.com/fourkitchens/emulsify/pull/268#issuecomment-402732950, I couldn't find any other instance of `components/css` in emulsify.php
